src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -291,65 +312,74 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
+            await UpdateStatusBarAsync(dte, description, cancellationToken).ConfigureAwait(false);
 
             try
             {
-                return this.ThreadingContext.JoinableTaskFactory.Run(() =>
+                return await this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
                 {
-                    return this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
+                    var installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
+                    if (installedPackagesMap.ContainsKey(packageName))
+                        return false;
+
+                    // Once we start the installation, we can't cancel anymore.
+                    cancellationToken = default;
+                    if (version == null)
                     {
-                        await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-
-                        var installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
-                        if (installedPackagesMap.ContainsKey(packageName))
-                            return false;
-
-                        dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_0, packageName);
-
-                        if (versionOpt == null)
-                        {
-                            _packageInstaller.Value.InstallLatestPackage(
-                                source, dteProject, packageName, includePrerelease, ignoreDependencies: false);
-                        }
-                        else
-                        {
-                            _packageInstaller.Value.InstallPackage(
-                                source, dteProject, packageName, versionOpt, ignoreDependencies: false);
-                        }
-
-                        installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
-                        var installedVersion = installedPackagesMap.TryGetValue(packageName, out var version) ? version : null;
-                        dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_0_completed,
-                            GetStatusBarText(packageName, installedVersion));
-
-                        return true;
-                    }, cancellationToken);
-                });
+                        _packageInstaller.Value.InstallLatestPackage(
+                            source, dteProject, packageName, includePrerelease, ignoreDependencies: false);
+                    }
+                    else
+                    {
+                        _packageInstaller.Value.InstallPackage(
+                            source, dteProject, packageName, version, ignoreDependencies: false);
```

Comment:
❗ This requires a preceding call to switch to the main thread until NuGet/Home#11021 is complete (see other comment on the subject)

Reply:
➡️ No change required here once NuGet/NuGet.Client#4209 merges.